File: week06/explore_phila_data/prepare_parquet_septa_gtfs.mjs
import * as csv from 'csv/sync';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import { parquetWriteFile } from 'hyparquet-writer';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const RAW_DATA_DIR = path.join(__dirname, 'raw_data/');
const PREPARED_DATA_DIR = path.join(__dirname, 'prepared_data/');

const GTFS_FEEDS = ['septa_bus', 'septa_rail'];

for (const gtfsFeed of GTFS_FEEDS) {
  const gtfsFeedFolder = path.join(RAW_DATA_DIR, gtfsFeed + '/');
  for (const gtfsFileName of await fs.readdir(gtfsFeedFolder)) {

    // Read the data from the raw GTFS CSV file.
    const gtfsFilePath = gtfsFeedFolder + gtfsFileName;
    const content = await fs.readFile(gtfsFilePath, 'utf8');
    const rows = csv.parse(content, { "columns": true, "skip_empty_lines": true });

    if (rows.length === 0) continue;

    // Build column arrays
    const columnNames = Object.keys(rows[0]);
    const columns = Object.fromEntries(columnNames.map(name => [name, []]));
    for (const row of rows) {
      for (const name of columnNames) {
        columns[name].push(row[name]);
      }
    }

    // Write to Parquet
    const outputFolder = path.join(PREPARED_DATA_DIR, gtfsFeed + '/');
    const outputPath = path.join(outputFolder, gtfsFileName.replace('.txt', '.parquet'));
    await fs.mkdir(outputFolder, { recursive: true });

    parquetWriteFile({
      filename: outputPath,
      columnData: columnNames.map(name => ({ name, data: columns[name] })),
    });
  }
}

File: week06/explore_phila_data/prepare_parquet_septa_gtfs.py
import csv
import pathlib
import pyarrow as pa
import pyarrow.parquet as pq

RAW_DATA_DIR = pathlib.Path(__file__).parent / 'raw_data'
PREPARED_DATA_DIR = pathlib.Path(__file__).parent / 'prepared_data'

GTFS_FEEDS = ['septa_bus', 'septa_rail']

for gtfs_feed in GTFS_FEEDS:
    gtfs_feed_folder = RAW_DATA_DIR / gtfs_feed
    for gtfs_path in gtfs_feed_folder.iterdir():

        # Read the data from the raw GTFS CSV file.
        with gtfs_path.open('r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            rows = list(reader)

        if not rows:
            continue

        # Build column arrays
        column_names = list(rows[0].keys())
        columns = {name: [row[name] for row in rows] for name in column_names}

        # Write to Parquet
        output_folder = PREPARED_DATA_DIR / gtfs_feed
        output_path = output_folder / f'{gtfs_path.stem}.parquet'
        output_folder.mkdir(parents=True, exist_ok=True)

        table = pa.table({name: pa.array(values) for name, values in columns.items()})
        pq.write_table(table, output_path)
